### PR TITLE
Move guild cache to ETS table.

### PR DESCRIPTION
**Pull request by jchristgit**
This is a draft for early comments. #254 should be merged beforehand

### CODE REVIEW

**Comment by jchristgit** [comment]
oh, I forgot to mention: This is fully backwards-compatible.

**Comment by jchristgit** [comment]
@Kraigie

File: lib/nostrum/cache/guild_cache.ex [hunk 1]
```diff
@@ -173,4 +188,172 @@ defmodule Nostrum.Cache.GuildCache do
...
+    {old, new, new_members} = upsert(guild.members, member["user"]["id"], member, Member)
+    new_guild = %{guild | members: new_members}
+    true = :ets.update_element(@table_name, guild_id, {2, new_guild})
+    {old, new}
+  end
+
+  @doc false
+  def member_chunk(guild_id, member_chunk) do
+    # CHONK like craigs cats
+    # or are they? i don't remember
```
Comment (Kraigie) [inline lib/nostrum/cache/guild_cache.ex:280]: One of them is chonky, yes

File: lib/nostrum/cache/guild_cache.ex [hunk 2]
```diff
@@ -173,4 +188,172 @@ defmodule Nostrum.Cache.GuildCache do
...
+    true = :ets.update_element(@table_name, guild_id, {2, new_guild})
+    {old, new}
+  end
+
+  @doc false
+  def emoji_update(guild_id, emojis) do
+    [{_id, guild}] = :ets.lookup(@table_name, guild_id)
+    new = %{guild | emojis: emojis}
+    true = :ets.update_element(@table_name, guild_id, {2, new})
+    emojis
```
Comment (Kraigie) [inline lib/nostrum/cache/guild_cache.ex:247]: Did we check the return types on these? I took a quick spot check on this one, and for the emoji update event we send both the old and new lists.

**Review by Kraigie — COMMENTED** [review]
This LGTM.

Do you think we need to care about what happens when working with the ETS tables in the event there is a failure? Say we try and do some operation but it fails. It would bring the whole genstage process down.

File: lib/nostrum/cache/guild_cache.ex (hunk 1 above)
Comment (jchristgit) [reply lib/nostrum/cache/guild_cache.ex]: aab2015

File: lib/nostrum/cache/guild_cache.ex (hunk 2 above)
Comment (jchristgit) [reply lib/nostrum/cache/guild_cache.ex]: I'm mostly certain that I checked most return types. 68928905289fd9854e77b732223113093134a72e fixes the emojis.

Comment (jchristgit) [reply lib/nostrum/cache/guild_cache.ex]: I've rechecked the return types. It looks good to me. I've also added some typespecs.

**Comment by Th3-M4jor** [comment]
Any chance this will be merged soon?

I was told in the Discord channel that this PR fixes a KeyError, which if it happens more than once in the span of a couple minutes (Such as someone updating the color of more than one Role) the guild cache process won't be restarted again and the bot stops receiving events until it's manually restarted.

**Review by Kraigie — APPROVED** [review]
This LGTM. :shipit:

**Comment by jchristgit** [comment]
Thanks!

**Pull request merged**